Cleanup non-rest apiserver handlers

sttts opened the pull request:
- rename MuxHelper -> PathRecorderMux
- move non-rest handlers into routes packages within genericapiserver and `pkg/routes` (those from master)
- move ui and logs handlers out of genericapiserver (they are
  not generic)
- make version handler configurable (`config.EnableVersion`)

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32387)

<!-- Reviewable:end -->

Commit:
Cleanup non-rest apiserver handlers

- rename MuxHelper -> PathRecorderMux
- move non-rest handlers into routes packages within genericapiserver and
  `pkg/routes` (those from master)
- move ui and logs handlers out of genericapiserver (they are
  not generic)
- make version handler configurable (`config.EnableVersion`)

sttts commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -268,6 +269,8 @@ func Run(s *options.APIServer) error {
 		glog.Fatalf("Failed to initialize plugins: %v", err)
 	}
 
+	mux := http.NewServeMux()
```
I guess this initialization can go away, the default case for `RestfulContainer` should be enough.Trying to remember why I did this in the first place.

deads2k commented on pkg/genericapiserver/config.go:
```diff
@@ -318,16 +313,13 @@ func (c Config) New() (*GenericAPIServer, error) {
 	// Use CurlyRouter to be able to use regular expressions in paths. Regular expressions are required in paths for example for proxy (where the path is proxy/{kind}/{name}/{*})
 	s.HandlerContainer.Router(restful.CurlyRouter{})
-	s.MuxHelper = &apiserver.MuxHelper{Mux: s.mux, RegisteredPaths: []string{}}
+	s.Mux = apiserver.NewPathRecorderMux(c.RestfulContainer.ServeMux)
```
The if statement above suggests that `RestfulContainer` can be nil.

sttts replied:
good spot. this should be s.HandlerContainer.

sttts replied:
done

deads2k commented on pkg/genericapiserver/routes/version.go:
```diff
@@ -0,0 +1,29 @@
+type Version struct{}
+
+func (v Version) Install(mux *apiserver.PathRecorderMux, c *restful.Container) {
+	apiserver.InstallVersionHandler(mux, c)
```
Why isn't this inlined?

sttts replied:
done

deads2k commented on pkg/master/master.go:
```diff
@@ -179,16 +182,27 @@ func New(c *Config) (*Master, error) {
 		enableCoreControllers:   c.EnableCoreControllers,
 		deleteCollectionWorkers: c.DeleteCollectionWorkers,
 		tunneler:                c.Tunneler,
+		enableUISupport:         c.EnableUISupport,
```
Wasn't expecting this.  What do you need it for after you've already added them to the mux?

sttts replied:
Good point. There were some tests for them. That's the only use. Will remove them.

sttts replied:
done

deads2k commented:
Yes, this is what I had in mind.  I think its a big improvement.

deads2k commented:
@sttts you still adding to this?  I think I'd like stop adding to this pull, finalize what's here, and merge as the next refactoring step.

sttts commented:
@deads2k addressed your comments. That's all I would put in this PR. Much easier to proceed in smaller steps...

deads2k commented:
> @deads2k addressed your comments. That's all I would put in this PR. Much easier to proceed in smaller steps...

squash and I'll take a look on monday

sttts commented:
Squashed and rebased. gce tests seem to be in trouble right now.

deads2k commented on pkg/genericapiserver/config.go:
```diff
@@ -336,29 +328,28 @@ func (c Config) New() (*GenericAPIServer, error) {
-	if c.EnableUISupport {
-		ui.InstallSupport(s.MuxHelper, c.EnableSwaggerSupport && c.EnableSwaggerUI)
+	if c.EnableSwaggerSupport && c.EnableSwaggerUI {
+		routes.SwaggerUI{}.Install(s.Mux, s.HandlerContainer)
```
We need to followup and see if swagger ui is generic.  As I recall, it didn't actually work generically when I tried to embed it downstream.

nikhiljindal replied:
Curious to know what did not work when you tried it.
Should be generic. Reads the spec and displays it in a UI.

deads2k commented on pkg/genericapiserver/routes/version.go:
```diff
@@ -0,0 +1,50 @@
+type Version struct{}
+
+// InstallVersionHandler registers the APIServer's `/version` handler
+func (v Version) Install(mux *apiserver.PathRecorderMux, c *restful.Container) {
```
This isn't generic.

sttts replied:
The route setup can be generic, the actual version data isn't. Compare https://github.com/openshift/origin/blob/a58b6d87f531075ce9b9d3caffca66c00dfcab59/pkg/cmd/server/origin/master.go#L361-L361.

sttts replied:
We can factor that in a follow-up PR.

deads2k replied:
> The route setup can be generic, the actual version data isn't. Compare https://github.com/openshift/origin/blob/a58b6d87f531075ce9b9d3caffca66c00dfcab59/pkg/cmd/server/origin/master.go#L361-L361.

That's of dubious value.  Setting this endpoint for another server will break `kubectl` nearly 100% of the time and openshift wouldn't use this since we'd have different routes anyway.

I'm ok with the code for the route living in generic, but I don't think it should be called from the genericapiserver code.  Let the caller set it.

sttts replied:
Moved it to the master package. I think there it belong right now if is not generic.

deads2k commented on pkg/genericapiserver/routes/swaggerui.go:
```diff
@@ -0,0 +1,40 @@
+// SwaggerUI exposes files in third_party/swagger-ui/ under /swagger-ui.
+type SwaggerUI struct{}
+
+func (l SwaggerUI) Install(mux *apiserver.PathRecorderMux, c *restful.Container) {
```
If this points to something that's been bindata'd, it isn't generic

sttts replied:
Why not? If the source is static, it can be a generic apiserver feature. Haven't checked. Have you?

deads2k replied:
> Why not? If the source is static, it can be a generic apiserver feature. Haven't checked. Have you?

We can deal with it post-rebase, but I'm pretty certain this isn't generic.

deads2k commented on pkg/master/routes/metrics.go:
```diff
@@ -0,0 +1,53 @@
+// DefaultMetrics installs the default prometheus metrics handler
+type DefaultMetrics struct{}
+
+func (m DefaultMetrics) Install(mux *apiserver.PathRecorderMux, c *restful.Container) {
```
This one should be generic

deads2k replied:
> This one should be generic

I see this as outstanding.

deads2k commented:
Please add an integration test which proves that the endpoints you're installing into the mix end up behind the authn/authz filters.

sttts commented:
Added `TestNotRestRoutesHaveAuth` to genericapiserver_test.

sttts commented:
@deads2k ptal

deads2k commented:
Test looks fine.

deads2k commented:
lgtm, squash

sttts commented:
rebased.

nikhiljindal commented on federation/cmd/federation-apiserver/app/server.go:
```diff
@@ -38,6 +38,7 @@ import (
 	"k8s.io/kubernetes/pkg/genericapiserver/authorizer"
 	genericoptions "k8s.io/kubernetes/pkg/genericapiserver/options"
 	genericvalidation "k8s.io/kubernetes/pkg/genericapiserver/validation"
+	"k8s.io/kubernetes/pkg/master/routes"
```
We do not want federation to depend on pkg/master which pulls in all kubernetes APIs.
We moved the common things out of pkg/master into pkg/genericapiserver

nikhiljindal commented on federation/cmd/federation-apiserver/app/server.go:
```diff
@@ -196,6 +197,8 @@ func Run(s *options.ServerRunOptions) error {
 		return err
 	}
 
+	routes.Version{}.Install(m.Mux, m.HandlerContainer)
```
Version should be generic.
Without it `kubectl version` wont work.

sttts replied:
It's kube specific at least as it depends on the Kubernetes version package. We could move the whole handler into the version package. How does that sound? Another alternative would be to abstract the version getter into an interface and inject that via the config.

lavalamp replied:
This is just a drive by since I saw this in my email, but making the
generic apiserver take interfaces to reduce the import dependency tree
sounds great to me :)

On Tue, Sep 13, 2016 at 1:33 PM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In federation/cmd/federation-apiserver/app/server.go
> <https://github.com/kubernetes/kubernetes/pull/32387#discussion_r78638591>
> :
>
> > @@ -196,6 +197,8 @@ func Run(s *options.ServerRunOptions) error {
> >  		return err
> >  	}
> >
> > +	routes.Version{}.Install(m.Mux, m.HandlerContainer)
>
> It's kube specific at least as it depends on the Kubernetes version
> package. We could move the whole handler into the version package. How does
> that sound? Another alternative would be to abstract the version getter
> into an interface and inject that via the config.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32387/files/bd264d43d1c8c0aee12257bcee14754c7a69e761#r78638591>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAngljfVFh7YD0Vn3mZdqUOIkFI4MCG0ks5qpwipgaJpZM4J5JSU>
> .
>

nikhiljindal replied:
By kubernetes version package, do you mean `pkg/version`? We can move that to `genericapiserver/version` ? (There is no kubernetes specific code there, just some git references)

sttts replied:
Am still undecided. The version package `base.go` is customized via ldflags. So this could be done by any instantiation of the genericapiserver.

The actual interpretation of the version values in the kubelet is done not in the generic discovery client itself, but in kubelet code (compare https://github.com/kubernetes/kubernetes/pull/32493/files).

The discovery client assumes a certain json schema. So at least this must be generic. I think the best thing is to inject a version.Info object into the genericapiserver. In the case of kube-apiserver this can be set via ldflags. But other instances of genericapiserver might choose different ways to return a proper version.Info.

@deads2k what's your opinion?

nikhiljindal replied:
I assume you meant kubectl there and not kubelet?
Can you point to ldflags code you are talking about?
[base.go](https://github.com/kubernetes/kubernetes/blob/f2ddd60eb9e7e9e29f7a105a9a8fa020042e8e52/pkg/version/) just has strings

sttts replied:
kubectl of course, yes.

I referred to this comment, haven't investigated where the values are overwritten with actual values in the build system: https://github.com/sttts/kubernetes/blob/5ead89b5bb1849d59589a972047e2876be479e62/pkg/version/base.go#L22-L22

nikhiljindal commented on hack/build-ui.sh:
```diff
@@ -37,7 +37,7 @@ readonly SWAGGER_PKG="swagger"
   local pkg="$1"
   local src="$2"
-  local output_file="pkg/ui/data/${pkg}/datafile.go"
+  local output_file="pkg/genericapiserver/routes/data/${pkg}/datafile.go"
```
Is this intentional?

You moved UI out of genericapiserver, but moved the datafile for it inside genericapiserver?

sttts replied:
The data is for swagger, the ui you mean is the dashboard redirection. As these two are hardly related, I have split them.

nikhiljindal replied:
oh I was under the impression that this data file contains the dashboard files as well. Its only swagger UI files?

nikhiljindal replied:
fwiw, when this was first introduced, it only had the UI (dashboard) files

sttts replied:
The dashboard is launched via pod+service nowadays. But can very well be that this was the case before.

nikhiljindal commented on pkg/genericapiserver/config.go:
```diff
@@ -60,14 +60,11 @@ type Config struct {
 	EnableSwaggerSupport bool
 	// Allow downstream consumers to disable swagger ui.
-	// Note that this is ignored if either EnableSwaggerSupport or EnableUISupport is false.
+	// Note that this is ignored EnableSwaggerSupport is false
```
ignored if

sttts replied:
fixed

nikhiljindal commented on pkg/genericapiserver/config.go:
```diff
@@ -318,16 +313,13 @@ func (c Config) New() (*GenericAPIServer, error) {
 	}
 
 	if c.RestfulContainer != nil {
-		s.mux = c.RestfulContainer.ServeMux
```
So we never use c.RestfulContainer.ServeMux?
Doesnt this break anyone who is setting that and expecting it to be used.

I always assumed that its openshift that is setting ServeMux. Is it not?

sttts replied:
If `c.RestfulContainer` is set, it is used for `s.HandlerContainer`. Further down `s.Mux` is set to the `ServeMux` of the container. So it's not ignored. Any use case I am missing here?

nikhiljindal replied:
yes ok. You are right

nikhiljindal commented on pkg/master/master.go:
```diff
@@ -179,11 +179,21 @@ func New(c *Config) (*Master, error) {
 
-	s, err := c.Config.New()
+	gc := *c.Config                                              // copy before mutations
+	gc.EnableSwaggerUI = gc.EnableSwaggerUI && c.EnableUISupport // disable swagger UI if general UI supports is
```
is -> it

sttts replied:
fixed

nikhiljindal commented:
Have added a few comments.

What is the motivation for this change?
Do you have an apiserver in mind that does not want /version, /logs and /ui endpoints?

nikhiljindal commented:
cc @kubernetes/sig-api-machinery

sttts commented:
@nikhiljindal yes, it's about making the genericapiserver really minimal in the sense that it only provides routes around the actual api which is used by the clients. 

The /version endpoint isn't that clear (yet), I agree. It's only used in the discovery client AFAIK. The values under /version at least must be injected for any instantiation of the genericapiserver as they depend on the repo/project.

But `/log` and `/ui` are something you don't want in non-kube apiservers. E.g. think about apiservers sitting next to the kube apiserver serving custom api groups, with a client supporting federation making this transparent.

nikhiljindal commented:
> @nikhiljindal yes, it's about making the genericapiserver really minimal in the sense that it only provides routes around the actual api which is used by the clients.
> 

:+1: 

> The /version endpoint isn't that clear (yet), I agree. It's only used in the discovery client AFAIK. The values under /version at least must be injected for any instantiation of the genericapiserver as they depend on the repo/project.
> 

So do you want to let /version be there, then?
Am fine with removing /log and /ui.

> But /log and /ui are something you don't want in non-kube apiservers. E.g. think about apiservers sitting next to the kube apiserver serving custom api groups, with a client supporting federation making this transparent.

Yes the aim of starting this genericapiserver library was to make it easier for people to write apiserver to enable federated api servers. So I totally agree with you.

deads2k commented:
I'd really like to start by stripping everything down and building back up.  I don't know how common I think /version actually is.  OpenShift (the concrete example we have) won't host version information there.

How about we make the route easy to install (pass a strict through maybe?), but keep it out of the generic API server strict itself?  That will keep generic API server small and have it cleanly make use of the install pattern we're starting here?

nikhiljindal commented:
> I would say that genericapiserver is *only* about APIs.  There is a wrapper
> above which is "common API server conventions" which includes /metrics,
> /version, /healthz, etc.  But genericapiserver does not require those.

I agree that `/metrics`, `/logs`, `/ui`, `/swagger-ui` can be taken out of genericapiserver.
But I view `/healthz`, `/swagger-spec` and `/version` as special since kubectl expects those endpoints to exist. So I would like them to remain in genericapiserver.

Also, we still want a common place to put the endpoints that we are taking out of genericapiserver, to prevent duplicating code in federation and kube apiservers. pkg/master is not the right place. We do not want federation-apiserver to depend on pkg/master.

sttts commented:
Moved the version route back into the genericapiserver package, making it configurable with `Config.EnableVersion`.

@nikhiljindal I agree with the common package. How about `pkg/routes`?

Following @smarterclayton's comment, I have the following picture of where the packages should go:

- `pkg/master`, `federation`, OpenShift origin: instantiations of genericapiserver with concrete, installed api groups
- `pkg/genericapiserver`: a minimal http server, essentially the counterpart of what we support and expect in the client, but without the actual api groups being installed. Everything above that should be configurable (/, /version, /swaggerapi, /openapi, /swagger-ui).

  Rules of thumb:
  * everything that relates to the Kubernetes domain should live elsewhere (/ui, ssh tunnels)
  * everything that does not relate to providing an api should live else where (/metrics, /logs, /debug/pprof), preferably avoiding code duplication between kube-apiserver, federation-apiserver and OpenShift origin.
- `pkg/apiserver`: the kube-like rest api infrastructure, essentially implements the handlers which will serve api groups, discovery etc.

Some work for some follow-up PRs:
- move ssh tunneler out of genericapiserver
- move /debug/pprof route into a common routes packages

sttts commented:
@nikhiljindal I moved the master routes into pkg/routes and re-activated logs and ui redirection in the federation apiserver.

@nikhiljindal @deads2k ptal.

deads2k commented on pkg/genericapiserver/config.go:
```diff
@@ -71,6 +71,7 @@ type Config struct {
 	// allow downstream consumers to disable the index route
 	EnableIndex           bool
 	EnableProfiling       bool
+	EnableVersion         bool
```
Very frustrating since this wrong in every use-case except one.

@nikhiljindal what about making the `Version` route accept a struct and having the server that's actually using it provide the route?  We still wouldn't use it downstream, but at least it would clutter up the genericapiserver.

sttts replied:
Why is this wrong?

How generic is the discovery client supposed to be? That's the unanswered question here IMO. If it is not generic, let's move the version route to pkg/routes.

For reference: origin has `/version/openshift` (https://github.com/openshift/origin/blob/a58b6d87f531075ce9b9d3caffca66c00dfcab59/pkg/cmd/server/origin/master.go#L349-L349). 

Is `/version/openshift` used as a replacement of kube's /version? What will a kubectl used against an origin cluster see and use?

It feels pretty strange to have

deads2k replied:
> Is /version/openshift used as a replacement of kube's /version? What will a kubectl used against an origin cluster see and use?

When you change what `/version` returns, `--match-server-version` on `kubectl` goes crazy.  I'd be happy to eliminate that flag (we haven't found much use for it downstream), but its not a generic endpoint.

Even if it were, this should change to specify what you want served.

I'll settle for this to get the rest of the pull, but `Version` as it is today is not generic

sttts commented:
@k8s-bot test this issue: #30228

smarterclayton commented:
/healthz, /swagger-spec, and /version are what I would consider part of a
"generic server".  A generic API server does not require healthz (you can
use different paths for health), swagger spec (your discovery doc is
supposed to point to it), and version is informational at best.  So I
disagree that we should be mandating that all API servers have them except
by convention.

No client should fail because a server doesn't implement /version,
/healthz, or /swagger-spec at exactly those paths.

On Wed, Sep 14, 2016 at 3:43 AM, Nikhil Jindal <notifications@github.com>
wrote:

> I would say that genericapiserver is *only* about APIs. There is a wrapper
> above which is "common API server conventions" which includes /metrics,
> /version, /healthz, etc. But genericapiserver does not require those.
>
> I agree that /metrics, /logs, /ui, /swagger-ui can be taken out of
> genericapiserver.
> But I view /healthz, /swagger-spec and /version as special since kubectl
> expects those endpoints to exist. So I would like them to remain in
> genericapiserver.
>
> Also, we still want a common place to put the endpoints that we are taking
> out of genericapiserver, to prevent duplicating code in federation and kube
> apiservers. pkg/master is not the right place. We do not want
> federation-apiserver to depend on pkg/master.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32387#issuecomment-246932062>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p0j0Cg-1YvMBrtbixzO6PMUXszFYks5qp6WWgaJpZM4J5JSU>
> .
>

sttts commented:
@smarterclayton agreeing about that list. All of those are optional right now. With the exception of /healthz they are enabled via the Config, /healthz is created explicitly in master.go. I think the only left dissent is how and whether to enable/disable them. I propose to leave it like it is now in the PR. In a follow-up, I am planning to rework the mechanism to customize the handler chain(s). That will most probably have influence on the way some of the routes are set up (e.g. some of routes can only set up after apis are added).

deads2k commented:
> I think the only left dissent is how and whether to enable/disable them. I propose to leave it like it is now in the PR.

I like that.

smarterclayton commented:
i'm fine with now, I want to ensure we're discussing it elsewhere.  Adding
APIs to the required list for a client is itself adding an API and needs
buy in from a larger audience - we've gone too far in some directions
without considering the implications.

On Wed, Sep 14, 2016 at 11:10 AM, Kubernetes Bot <notifications@github.com>
wrote:

> GCE e2e build/test *passed* for commit b106358
> <https://github.com/kubernetes/kubernetes/commit/b1063583eb836f92d9d45209e59a91117d2b6a2f>
> .
>
>    - Test Results
>    <https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/32387/kubernetes-pull-build-test-e2e-gce/58830>
>    - Build Log
>    <http://pr-test.k8s.io/32387/kubernetes-pull-build-test-e2e-gce/58830/build-log.txt>
>    - Test Artifacts
>    <https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/32387/kubernetes-pull-build-test-e2e-gce/58830/artifacts/>
>    - Internal Jenkins Results
>    <http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//58830>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32387#issuecomment-247045493>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p7RRcvcsY_0EsX_lSA_ccSvqpiKcks5qqA5qgaJpZM4J5JSU>
> .
>

sttts commented:
For reference: kubectl works mostly fine without `/version`. But a number of commands (apply, create, convert, ....) are wired to this:

https://github.com/kubernetes/kubernetes/blob/519c13745f8c5cc3fa50897575fe7aaca04cfbf1/pkg/kubectl/cmd/util/factory.go#L749-L749

A 404 on /version let's the kubectl to fail hard.

smarterclayton commented:
Can you open a bug on that - we have users who want to disable access to
version because it allows targeted security attacks.  /version should not
be required for proper functioning of client unless the strict version
check is on.

On Wed, Sep 14, 2016 at 12:19 PM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> For reference: kubectl works mostly fine without /version. But a number
> of commands (apply, create, convert, ....) are wired to this:
>
> https://github.com/kubernetes/kubernetes/blob/
> 519c13745f8c5cc3fa50897575fe7aaca04cfbf1/pkg/kubectl/cmd/
> util/factory.go#L749-L749
>
> A 404 on /version let's the kubectl to fail hard.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32387#issuecomment-247068985>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p-4gMiuR4Tn21ltsrBY24P3mHPuMks5qqB6egaJpZM4J5JSU>
> .
>

lavalamp commented:
I agree that generic apiserver should be doing the minimal set of things. I'm fine with moving version and healthz elsewhere.

In practice I think we're going to require swagger (or how else is the discovery merger going to function?) so I'd be more hesitant about moving that out, but it's not that big of a deal either way.

(also Clayton you missed a "not" in a previous comment which makes it confusing!)

lavalamp commented on pkg/genericapiserver/routes/addon.go:
```diff
@@ -0,0 +1,29 @@
+import (
+	"github.com/emicklei/go-restful"
+
+	"k8s.io/kubernetes/pkg/apiserver"
```
generic apiserver should not import apiserver, no?

nikhiljindal replied:
Yes it should not, but we havent fixed that yet: https://github.com/kubernetes/kubernetes/issues/22103

sttts replied:
@lavalamp compare https://github.com/kubernetes/kubernetes/issues/22103#issuecomment-247281269 and https://github.com/kubernetes/kubernetes/pull/32387#issuecomment-246981162.

smarterclayton commented:
I think swagger is required, I just don't think it has to be at
/swagger-spec.  We had the path spec'd in the discovery proposal (so
discovery finds swagger, rather than vice versa).

On Wed, Sep 14, 2016 at 6:19 PM, Daniel Smith <notifications@github.com>
wrote:

> I agree that generic apiserver should be doing the minimal set of things.
> I'm fine with moving version and healthz elsewhere.
>
> In practice I think we're going to require swagger (or how else is the
> discovery merger going to function?) so I'd be more hesitant about moving
> that out, but it's not that big of a deal either way.
>
> (also Clayton you missed a "not" in a previous comment which makes it
> confusing!)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32387#issuecomment-247174507>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p9u7yI8YsMBPWcxPXrGF1OHEaZkUks5qqHLbgaJpZM4J5JSU>
> .
>

lavalamp commented:
I'm not sure what we gain by making the swagger path customizable...

On Wed, Sep 14, 2016 at 3:22 PM, Clayton Coleman <notifications@github.com>
wrote:

> I think swagger is required, I just don't think it has to be at
> /swagger-spec. We had the path spec'd in the discovery proposal (so
> discovery finds swagger, rather than vice versa).
>
> On Wed, Sep 14, 2016 at 6:19 PM, Daniel Smith <notifications@github.com>
> wrote:
>
> > I agree that generic apiserver should be doing the minimal set of things.
> > I'm fine with moving version and healthz elsewhere.
> >
> > In practice I think we're going to require swagger (or how else is the
> > discovery merger going to function?) so I'd be more hesitant about moving
> > that out, but it's not that big of a deal either way.
> >
> > (also Clayton you missed a "not" in a previous comment which makes it
> > confusing!)
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/kubernetes/kubernetes/pull/
> 32387#issuecomment-247174507>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/ABG_
> p9u7yI8YsMBPWcxPXrGF1OHEaZkUks5qqHLbgaJpZM4J5JSU>
> > .
> >
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32387#issuecomment-247175275>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglp0EWx6W8nxlmhRd2ko3tMp6QMk4ks5qqHOcgaJpZM4J5JSU>
> .
>

smarterclayton commented:
How will a summarizer present a unified swagger doc?  A summarizer should
appear to be a valid endpoint, but won't necessarily have a swagger spec
that makes sense.  Groups (and maybe group versions) have swagger docs, and
servers may be composites of swagger docs.  Assuming there is only one
swagger doc at any public API endpoint is already broken (for openshift)
and would be broken for a summarizer.  So I'd rather not bake that
assumption in at this level - I shouldn't have to make a call to an API
server hoping the swagger spec is at a URL - that's what discovery is for.
If we have discovery let's use it.

I won't go to the mat on this but I don't like having more than one magical
assumed url.

On Sep 14, 2016, at 6:42 PM, Daniel Smith <notifications@github.com> wrote:

I'm not sure what we gain by making the swagger path customizable...

On Wed, Sep 14, 2016 at 3:22 PM, Clayton Coleman <notifications@github.com>
wrote:

> I think swagger is required, I just don't think it has to be at
> /swagger-spec. We had the path spec'd in the discovery proposal (so
> discovery finds swagger, rather than vice versa).
>
> On Wed, Sep 14, 2016 at 6:19 PM, Daniel Smith <notifications@github.com>
> wrote:
>
> > I agree that generic apiserver should be doing the minimal set of
things.
> > I'm fine with moving version and healthz elsewhere.
> >
> > In practice I think we're going to require swagger (or how else is the
> > discovery merger going to function?) so I'd be more hesitant about
moving
> > that out, but it's not that big of a deal either way.
> >
> > (also Clayton you missed a "not" in a previous comment which makes it
> > confusing!)
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/kubernetes/kubernetes/pull/
> 32387#issuecomment-247174507>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/ABG_
> p9u7yI8YsMBPWcxPXrGF1OHEaZkUks5qqHLbgaJpZM4J5JSU>
> > .
> >
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <
https://github.com/kubernetes/kubernetes/pull/32387#issuecomment-247175275>,
> or mute the thread
> <
https://github.com/notifications/unsubscribe-auth/AAnglp0EWx6W8nxlmhRd2ko3tMp6QMk4ks5qqHOcgaJpZM4J5JSU
>
> .
>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/32387#issuecomment-247179340>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_pwMEHmKbQDL9SjeMKmGsVNl2bx-rks5qqHg6gaJpZM4J5JSU>
.

sttts commented:
Rebased and squashed.

@deads2k @lavalamp @nikhiljindal @smarterclayton anything left that must be done in this PR? Otherwise, let's continue iterating on genericapiserver in follow-up PRs.

Do we have an epic issue to track everything related to this endeavor already? If not, I will create one.

sttts commented:
I have created a tracking issue for all thing around making genericapiserver really generic: https://github.com/kubernetes/kubernetes/issues/32792

deads2k commented:
genericapiserver is already depending on apiserver.  This pull isn't making that worse and I'd say that its other changes are improvements on what we have.

@sttts is there anything else you want to include here?  If not, this lgtm.

lavalamp commented:
> How will a summarizer present a unified swagger doc?  A summarizer should
appear to be a valid endpoint, but won't necessarily have a swagger spec
that makes sense.  Groups (and maybe group versions) have swagger docs, and
servers may be composites of swagger docs. 

Yeah, this is a good question. @mbohlool already has code in that will serve an individual swagger doc for each group (IIRC). The main swagger doc should be mergeable as long as no two apiservers try to serve the same URL, AND as long as any duplicate type definitions are exactly duplicates. I think. I think some client tools may expect a single unified swagger document, so we probably shouldn't rule it out completely. @mbohlool-- do you have any thoughts?

deads2k commented:
> Yeah, this is a good question. @mbohlool already has code in that will serve an individual swagger doc for each group (IIRC). 

I think its likely that multiple summarized API server will server the same endpoint, they just won't want to federate it.  I'm not sure whether I'm agreeing with what you've said or not, but consider the case of something like a `certificates.k8s.io` server.  It may still have inspection API resources like the tokenreview or authorizationreview endpoints.  Those would be valid for its server and would allow it to play nice with `kubectl` client-side tools.  

The `certificates.k8s.io` server will want to federate its API group, but not every API group that it serves (not `authencation.k8s.io` as a for instance).

nikhiljindal commented:
https://github.com/kubernetes/kubernetes/issues/20935 has related prior discussion on merging swagger specs from multiple api servers

smarterclayton commented:
If we did a rolling upgrade that moved an API from one group to another
(i.e. we split storage / etc off into their own group) that would be an
upgrade where some servers had both API groups at the same time.  Would
they have to do the move, then do a binary upgrade separately?  Also, I
would assume that the API group cutover needed to be relatively atomic in
the presence of a summarizer, but during a rolling update different
backends would serve different swagger docs.

I think there are conveniences in a single swagger file for sure.

On Thu, Sep 15, 2016 at 4:24 PM, Daniel Smith <notifications@github.com>
wrote:

> How will a summarizer present a unified swagger doc? A summarizer should
> appear to be a valid endpoint, but won't necessarily have a swagger spec
> that makes sense. Groups (and maybe group versions) have swagger docs, and
> servers may be composites of swagger docs.
>
> Yeah, this is a good question. @mbohlool <https://github.com/mbohlool>
> already has code in that will serve an individual swagger doc for each
> group (IIRC). The main swagger doc should be mergeable as long as no two
> apiservers try to serve the same URL, AND as long as any duplicate type
> definitions are exactly duplicates. I think. I think some client tools may
> expect a single unified swagger document, so we probably shouldn't rule it
> out completely. @mbohlool-- do you have any thoughts?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32387#issuecomment-247443872>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pz42Wsxq8b_mdMFN-gKxK9BvyjPDks5qqamZgaJpZM4J5JSU>
> .
>

sttts commented:
@deads2k a collection of possible follow-up actions is at https://github.com/kubernetes/kubernetes/issues/32792. This PR is complete. Applying lgtm.

lavalamp commented:
Yeah, I think the summarizer would go by its config file as two which group
should be served from which apiserver. It should delete from an apiserver's
swagger spec anything that it's not actually serving for. I think that
mostly takes care of the migration case (both apiservers can serve the
group, then you flip over the summarizer, then you can take the group out
of the old apiserver).

On Fri, Sep 16, 2016 at 7:51 AM, Clayton Coleman <notifications@github.com>
wrote:

> If we did a rolling upgrade that moved an API from one group to another
> (i.e. we split storage / etc off into their own group) that would be an
> upgrade where some servers had both API groups at the same time. Would
> they have to do the move, then do a binary upgrade separately? Also, I
> would assume that the API group cutover needed to be relatively atomic in
> the presence of a summarizer, but during a rolling update different
> backends would serve different swagger docs.
>
> I think there are conveniences in a single swagger file for sure.
>
> On Thu, Sep 15, 2016 at 4:24 PM, Daniel Smith <notifications@github.com>
> wrote:
>
> > How will a summarizer present a unified swagger doc? A summarizer should
> > appear to be a valid endpoint, but won't necessarily have a swagger spec
> > that makes sense. Groups (and maybe group versions) have swagger docs,
> and
> > servers may be composites of swagger docs.
> >
> > Yeah, this is a good question. @mbohlool <https://github.com/mbohlool>
> > already has code in that will serve an individual swagger doc for each
> > group (IIRC). The main swagger doc should be mergeable as long as no two
> > apiservers try to serve the same URL, AND as long as any duplicate type
> > definitions are exactly duplicates. I think. I think some client tools
> may
> > expect a single unified swagger document, so we probably shouldn't rule
> it
> > out completely. @mbohlool-- do you have any thoughts?
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/kubernetes/kubernetes/pull/
> 32387#issuecomment-247443872>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/ABG_pz42Wsxq8b_mdMFN-
> gKxK9BvyjPDks5qqamZgaJpZM4J5JSU>
> > .
>
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/32387#issuecomment-247620826>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAngluPl2I4c_jW3qcWqMtL2z2dCuYbZks5qqqzggaJpZM4J5JSU>
> .
>

smarterclayton commented:
I'm thinking of even subtler stuff - models that have changed, a refactor
of swagger output for various reasons, etc

On Sep 16, 2016, at 12:58 PM, Daniel Smith <notifications@github.com> wrote:

Yeah, I think the summarizer would go by its config file as two which group
should be served from which apiserver. It should delete from an apiserver's
swagger spec anything that it's not actually serving for. I think that
mostly takes care of the migration case (both apiservers can serve the
group, then you flip over the summarizer, then you can take the group out
of the old apiserver).

On Fri, Sep 16, 2016 at 7:51 AM, Clayton Coleman <notifications@github.com>
wrote:

> If we did a rolling upgrade that moved an API from one group to another
> (i.e. we split storage / etc off into their own group) that would be an
> upgrade where some servers had both API groups at the same time. Would
> they have to do the move, then do a binary upgrade separately? Also, I
> would assume that the API group cutover needed to be relatively atomic in
> the presence of a summarizer, but during a rolling update different
> backends would serve different swagger docs.
>
> I think there are conveniences in a single swagger file for sure.
>
> On Thu, Sep 15, 2016 at 4:24 PM, Daniel Smith <notifications@github.com>
> wrote:
>
> > How will a summarizer present a unified swagger doc? A summarizer should
> > appear to be a valid endpoint, but won't necessarily have a swagger spec
> > that makes sense. Groups (and maybe group versions) have swagger docs,
> and
> > servers may be composites of swagger docs.
> >
> > Yeah, this is a good question. @mbohlool <https://github.com/mbohlool>
> > already has code in that will serve an individual swagger doc for each
> > group (IIRC). The main swagger doc should be mergeable as long as no two
> > apiservers try to serve the same URL, AND as long as any duplicate type
> > definitions are exactly duplicates. I think. I think some client tools
> may
> > expect a single unified swagger document, so we probably shouldn't rule
> it
> > out completely. @mbohlool-- do you have any thoughts?
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/kubernetes/kubernetes/pull/
> 32387#issuecomment-247443872>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/ABG_pz42Wsxq8b_mdMFN-
> gKxK9BvyjPDks5qqamZgaJpZM4J5JSU>
> > .
>
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <
https://github.com/kubernetes/kubernetes/pull/32387#issuecomment-247620826>,
> or mute the thread
> <
https://github.com/notifications/unsubscribe-auth/AAngluPl2I4c_jW3qcWqMtL2z2dCuYbZks5qqqzggaJpZM4J5JSU
>
> .
>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/32387#issuecomment-247652327>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p8YTiSJJbTYVI5u-eH8o1RUM7WkUks5qqsqWgaJpZM4J5JSU>
.

nikhiljindal commented:
> Yeah, I think the summarizer would go by its config file as two which group
> should be served from which apiserver. It should delete from an apiserver's
> swagger spec anything that it's not actually serving for. I think that
> mostly takes care of the migration case (both apiservers can serve the
> group, then you flip over the summarizer, then you can take the group out
> of the old apiserver).

Adding each group version to the config file seems too much work for the admin. Anytime a new group version is added to a server, config update will be required. The current model, where discovery summarizer automatically discovers the new group version is better.

Another alternative is to use the server registration order as the priority order. If a group version is being served by 2 apiservers, the server that was registered before is the preferred one. To move a group version from one server to another, we first add it to the new server and then remove it from the first. No updates required to the discovery summarizer.

deads2k commented:
> Adding each group version to the config file seems too much work for the admin.

I had assumed that we'd be trying to avoid having any config file updates at all.  We could use a cluster-scoped resource to track information like:

```go
type FederatedAPIGroupVersion struct{
    ObjectMeta // name is forced to version.group

    Servers []string
    Priority int // priority ordering for serving discovery

    Status Status // this would have the information included in discovery along with a summary of /healthz
}
```

This would allow a single server to host multiple group versions, while still allowing different versions to be hosted by different API servers/services.  It would also allow a server to host a group version without exposing it.  The priority would ensure that the cluster-admin is in control of which resources have priority by default on his systems (kubectl respects the order returned via discovery).
